apps/emqx_bridge_azure_blob_storage/src/emqx_bridge_azure_blob_storage_connector.erl

```diff
@@ -0,0 +1,763 @@
+        {error, Reason} ->
+            {error, Reason}
+    end;
+process_write(TransferState = #{started := true, buffer_size := 0}) ->
+    {ok, TransferState};
+process_write(TransferState0 = #{started := true}) ->
+    #{
+        buffer := Buffer,
+        buffer_size := BufferSize,
+        max_block_size := MaxBlockSize
+    } = TransferState0,
+    case BufferSize > MaxBlockSize of
+        true ->
+            {IOData, NewBuffer} = take_chunk(Buffer, MaxBlockSize),
+            ?tp(azure_blob_storage_will_write_chunk, #{}),
+            do_process_write(IOData, NewBuffer, TransferState0);
+        false ->
+            NewBuffer = [],
+            do_process_write(Buffer, NewBuffer, TransferState0)
+    end.
+
+do_process_write(IOData, NewBuffer, TransferState0 = #{started := true}) ->
+    #{
+        blob := Blob,
+        container := Container,
+        num_blocks := NumBlocks,
+        pool := Pool
+    } = TransferState0,
+    case append_data(Pool, Container, Blob, block_id(NumBlocks), IOData) of
+        {ok, _} ->
+            BufferSize = iolist_size(NewBuffer),
+            TransferState = TransferState0#{
+                buffer := NewBuffer,
+                buffer_size := BufferSize,
+                num_blocks := NumBlocks + 1
+            },
+            process_write(TransferState);
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec process_complete(transfer_state()) ->
+    {ok, term()}.
+process_complete(TransferState) ->
+    #{
+        blob := Blob,
+        container := Container,
+        num_blocks := NumBlocks,
+        pool := Pool
+    } = TransferState,
+    BlockRefs = [{block_id(N), latest} || N <- lists:seq(0, NumBlocks - 1)],
+    case put_block_list(Pool, Container, Blob, BlockRefs) of
+        {ok, _} ->
+            {ok, #{num_blocks => NumBlocks}};
+        {error, Reason} ->
+            exit({upload_failed, Reason})
+    end.
+
+%%------------------------------------------------------------------------------
+%% `emqx_template' API
+%%------------------------------------------------------------------------------
+
+-spec lookup(emqx_template:accessor(), {_Name, buffer_map()}) ->
+    {ok, integer() | string()} | {error, undefined}.
+lookup([<<"action">>], {ActionName, _BufferMap}) ->
+    {ok, mk_fs_safe_string(ActionName)};
+lookup(Accessor, {_ActionName, BufferMap = #{}}) ->
+    lookup_buffer_var(Accessor, BufferMap);
+lookup(_Accessor, _Context) ->
+    {error, undefined}.
+
+lookup_buffer_var([<<"datetime">>, Format], #{since := Since}) ->
+    {ok, format_timestamp(Since, Format)};
+lookup_buffer_var([<<"datetime_until">>, Format], #{until := Until}) ->
+    {ok, format_timestamp(Until, Format)};
+lookup_buffer_var([<<"sequence">>], #{seq := Seq}) ->
+    {ok, Seq};
+lookup_buffer_var([<<"node">>], #{}) ->
+    {ok, mk_fs_safe_string(atom_to_binary(erlang:node()))};
+lookup_buffer_var(_Binding, _Context) ->
+    {error, undefined}.
```

Comment:
Might be a good time to introduce something like a small `emqx_connector_aggreg_buffer_ctx` module, that implements `emqx_template` behaviour for `#buffer{}`s, and put this stuff (except for `<<"node">>`) there? Though we'd have to keep `mk_fs_safe_string/1` on this level. Not the prettiest solution, but could help minimize code duplication and the need for that weird `buffer_map()` type.

Reply:
:heavy_check_mark: